20220714-lookup-api.md

```diff
@@ -0,0 +1,460 @@
+# Meta
+[meta]: #meta
+- **Name**: Lookup API
+- **Start Date**: 2022-07-14
+- **Author(s)**: jon-whit
+- **Status**: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- **RFC Pull Request**: (leave blank)
+- **Relevant Issues**:
+  - https://github.com/openfga/openfga/issues/101
+  - https://github.com/openfga/api/issues/10
+- **Supersedes**: N/A
+
+# Summary
+[summary]: #summary
+
+In some scenarios like UI filtering or auditing, developers need to answer queries like 
+
+> list all documents that user:X can read
+
+More specifically, given an object type, relation, and user, list all of the object ids of the given object type the user has the provided relationship with.
+
+Depending on the use case, this problem can be solved with the existing Check API, as described in our [Search with Permissions](https://openfga.dev/docs/interacting/search-with-permissions) documentation, but in some cases it is more practical to have a dedicated API to serve this functionality.
+
+The Lookup API addresses [Search with Permissions (Option 3)](https://openfga.dev/docs/interacting/search-with-permissions#option-3-build-a-list-of-ids-then-search). It is specifically targeting the Search with Permissions use case for small object collections (on the order of thousands). For Search with Permissions on larger collections (tens of thousands and beyond) an OpenFGA client should build an external index via the ReadChanges (e.g. Watch) API or, in the future, via the OpenFGA Indexer.
```

Comment:
The scope of this proposal discusses the Lookup API, which is designed to "lookup" all of the objects that a user has a particular relationship too. However, in the near future we may also have the reverse form of this API. That is, given an object I want to "lookup" all of the users/subjects (in their expanded form) that have a particular relationship with a single object. LookupObjects would return a list of object ids while LookupUsers would return a list of user ids (for example).

It may be worth naming this API something like `LookupObjects` and the other `LookupUsers` to allow for this growth in the API definitions? WDYT?

Reply:
Yes, I prefer LookupObjects.

Reply:
BTW, what if we use `List` instead of `Lookup`? 

The term seems more commonly used in API definitions than lookup (e.g. google for "api ListUsers")

Reply:
> 

That sounds good. Let's go that route. `List` is just as good as `Lookup` and it's used more unanimously.

Reply:
@aaguiarz I've refactor this proposal to reflect these changes discussed. Now it's `ListObjects` instead of `LookupObjects`.